Add bulk claiming at /on/npm/

whit537 opened the pull request:
Picking up from #4416.

### Todo

- [x] move package API to a mixin: #4495
- [x] insert into `emails` during `make_participant`: #4495
- [x] add `get_packages_for_claiming`
- [x] modify `get_packages_for_claiming` to sort/group as desired
- [x] link to the email settings page
- [x] set the avatar as the npm logo
- [x] disable apply when there are no claimable packages
- [x] indicate when a package is already claimed, differentiating claimed-by-self and -by-other
- [x] fix test suite
- [x] wire up to `emails/modify.json`, send to address displayed
- [x] verify that we test what happens when someone tries to claim an already claimed package
- [x] ~~reuse open review ticket where possible~~—bumping to #4505
- [x] tie into review process better

whit537 commented:
Working up a ttw test.

![screen shot 2017-05-31 at 5 07 01 am](https://cloud.githubusercontent.com/assets/134455/26624482/0111213e-45bf-11e7-8832-dc4385a25c29.png)

whit537 commented:
I'm working on this commit that actually adds the email address during `make_participant` (instead of just blindly inserting it). It's a simple change (fc36e0c5840f2854019075c2714949f67212eed1) but has follow-on effects for the rest of the test suite. I'm down to one more failure, and it's weird. It's in `test_teams.py`, and it's an interaction between several tests, with the failure manifesting in `test_error_message_for_slug_collision`. It's a failure to overwrite a VCR file—but! it only shows up if five other tests are present. Of the six I've got with it right now (a3872a50ed7f971bf8d37acc32795c04a97872a0) if any of the other five are gone the test passes. If all other six are present then it fails. 🐨

whit537 commented:
Heck with it. Let's just rerecord the VCR fixture. 😉

whit537 commented:
Moved off this PR into #4495.

whit537 commented:
![screen shot 2017-06-02 at 11 23 46 am](https://cloud.githubusercontent.com/assets/134455/26732544/f3c70a16-4785-11e7-9b07-e3714b499925.png)

whit537 commented:
Rebased on master, was 4587ba8a3e2989d385d5a6c91a941e5e09378ce5.

whit537 commented:
Bringing over some local notes:

```
think through what happens when I put someone else's email in there                                           
    they get a link that's like "Claim your packages on Gratipay!"                                            
    they think, "Um, okay, why not?"                                                                          
        they think that we sent it to them                                                                    
    they click                                                                                                
    they are now sharing an account with an attacker                                                          
                                                                                                              
    I think we need to confirm again on landing                                                               
        "You are linking to thus and such account"                                                            
                                                                                                              
    this is a general problem with email verification, not specific here                                      
                                                                                                              
    At the very least we need to have the message in the email be "Sound familiar?"                           
        Try to communicate that they supposedly initiated the action
```

whit537 commented:
![screen shot 2017-06-05 at 11 33 37 am](https://cloud.githubusercontent.com/assets/134455/26790748/d682bf5e-49e2-11e7-8dca-c6541a1ff264.png)

whit537 commented:
![screen shot 2017-06-06 at 5 18 16 pm](https://user-images.githubusercontent.com/134455/26852389-25f49568-4adc-11e7-9e00-7032eec6aa8f.png)

whit537 commented:
Ready for review, @rohitpaulk! Commits are a bit sprawling, pushing the limit on max lines delta. Lmk if I should factor out a PR for ease.

whit537 commented:
Rebased, was c95af1e58150771db85f6772def8a8df07950f5b.

mattbk commented:
It all seems to work for me, but (as a new user), what do I do once I'm here? What's the next step?

![image](https://user-images.githubusercontent.com/3729038/26946963-a0c755e8-4c56-11e7-86b3-ab6182098355.png)

mattbk commented:
I like the automatic PayPal route creation if a person doesn't have the PayPal route set up yet. :+1:

![image](https://user-images.githubusercontent.com/3729038/26950080-1336f650-4c62-11e7-98da-01762cf3ced1.png)

whit537 commented:
Rebased (was 5d1e8095a152757341b50b3a4ea66f30b0650533) and added 51172012b4c8e07d21df90d010803bde8d598554:

![screen shot 2017-06-12 at 10 52 42 am](https://user-images.githubusercontent.com/134455/27039877-4d8927ee-4f5d-11e7-831e-e2af0828dc54.png)

whit537 commented:
e18babf4afafae14f94bc930efcc6a3f02d2dd37:

```
------=_Part_1298180_670291361.1497279406886
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 7bit

Greetings!

Thanks for applying to use Gratipay! The next step is for us to review your application, which happens on this public GitHub issue:

some-github-issue

You can watch and participate in our review there, and we will email you again when we finish (usually in about a week). Thanks again!

Something not right? Reply to this email for help.

----

Sent by Gratipay, LLC, https://gratipay.com/
716 Park Road, Ambridge, PA, 15003, USA
------=_Part_1298180_670291361.1497279406886
```

whit537 commented:
Better now, @mattbk? I ended up simplifying the `project-review.spt` template to allow reuse when there are multiple teams (by not referring to teams at all).

whit537 commented:
https://travis-ci.org/gratipay/gratipay.com/builds/242054280 hit #4511. Restarting ...

whit537 commented:
Travis found some regressions. I think 585f16e83f38bdec094d7f514ae069b62ca717a1 is clean?

whit537 commented:
Module 05a1295195fdff005758c9778aaa9ad7de572dbe (i.e., unrelated test imp that gets squished in the next PR, #4507), yes.

whit537 commented:
Rebased onto #4511 (which now includes 05a1295195fdff005758c9778aaa9ad7de572dbe), let's get that in first. Was 585f16e83f38bdec094d7f514ae069b62ca717a1.

mattbk commented:
If we're only sending one email, looks good to me. #4505 seems like the logical next step.

whit537 commented:
Ya, that and https://github.com/gratipay/gratipay.com/issues/4512.

rohitpaulk commented:
Rebased. Was 60d0e85f7a6ca669c74697c13b8ba2dd9ea653cb, now 486158949bce0b9d30c2b3880304a6d74b2c04ca

rohitpaulk commented on gratipay/models/participant/packages.py:
```diff
@@ -6,6 +6,61 @@
+        :param string manager: the name of the package manager on which to look
+            for potential packages
+
+        :return: a list of (:py:class:`~gratipay.models.package.Package`,
```
This doc looks outdated - looks like we're returning tuples with 4 elements, and not 3. The `is_primary` element seems to be missing.

rohitpaulk commented on gratipay/project_review_process.py:
```diff
@@ -31,12 +42,31 @@ def create_issue(self, *teams):
+        self.db.run("UPDATE teams SET review_url=%s WHERE id = ANY(%s)", (review_url, team_ids))
+        [team.set_attributes(review_url=review_url) for team in teams]
+
+        owner = Participant.from_username(owner_usernames.pop())
```
Might be better to use IDs instead of usernames for this (asserting we only have one unique value in the list + constructing a participant object)?

rohitpaulk replied:
Oh, we don't have IDs on the teams table. Ignore.

whit537 replied:
Yeah, #835.

rohitpaulk commented on gratipay/project_review_process.py:
```diff
@@ -73,6 +105,8 @@ def __init__(self, fp=sys.stdout):
         self.fp = fp
 
     def post(self, data):
+        """POST data to nowhere and return a URL of lies.
```
😃

rohitpaulk commented:
This is huge, going to try to split out some stuff...

rohitpaulk commented:
`get_packages_for_claiming` was modified in three commits - I extracted those changes out into #4513. Going to clean up the docs (https://github.com/gratipay/gratipay.com/pull/4488#discussion_r121636690) in that PR.

whit537 commented:
#4515 factored out, was a42f0023dc76052f268ad5493537b4b3403b3fa5.

whit537 commented:
Also factored out #4516 and #4517. More manageable now, @rohitpaulk? :-)

whit537 commented:
Rebased, was 6fc28cc0829b86460224cdd9dfcc480327133032.

rohitpaulk commented on www/on/npm/index.html.spt:
```diff
@@ -1,9 +1,129 @@
+{% else %}
+
+    {% if not packages_for_claiming %}
+    <p class="sorry">{{ _("No packages found.") }}</p>
```
Might be worth adding a note saying that we searched using the email addresses a, b, and c - and to update on npm if needed (just like the page for a single package)

rohitpaulk replied:
Oh wait, that is present below. Ignore.

rohitpaulk commented on www/on/npm/index.html.spt:
```diff
@@ -1,9 +1,129 @@
+banner = manager = "npm"
 suppress_sidebar = True
-website.redirect('/')  # nothing here yet
+npackages, Npackages = website.db.one('''
```
I think the naming here could be better. These values are only being pulled so that we can show the `x packages out of y are on Gratipay` - might be better to keep them in a single dict so that a confusion doesn't arise between `npackages`, `Npackages` and `packages_for_claiming`. 

How about - ```npm_stats = website.db.one('''select () as total_packages, () as claimed_packages''', return_as=dict)```, followed by ```n=npm_stats.total_packages, N=npm_stats.claimed_packages``` (in the template)?

whit537 replied:
Done in abfef42b687522c60b4061c6fa362fa6266e9bc0!

rohitpaulk commented on www/~/%username/emails/modify.json.spt:
```diff
@@ -46,7 +47,10 @@ if action in ('add-email', 'resend', 'start-verification'):
 
     participant.start_email_verification(address, *packages)
-    msg = _("Check your inbox for a verification link.")
+    if show_address_in_message:
```
Hmm, we always have an `address` - why not show it in the link?

rohitpaulk replied:
i.e. Why do we have to say `Check your inbox` when we have an `address` available

rohitpaulk replied:
Is it because we send multiple emails to addresses?

rohitpaulk replied:
I only see two emails being sent in `start_email_verification` - one is a verification notice to the primary email, the other is the actual verification link to the address being verified.

whit537 replied:
Because it's a content spoofing vector. I thought we had a HackerOne ticket about this but it's not jumping out at me on http://inside.gratipay.com/appendices/disclosures.

rohitpaulk replied:
Hmm, we allow reverifying email addresses? I see `new_email_verified=email in verified_emails` n `start_email_verification`. Digging deeper..

whit537 replied:
It's safe to show when the user is in control, but better not to for e.g. verifications, where a third party can initiate a message to any address.

rohitpaulk replied:
This is confusing 😕 `verification.spt` is used both to verify emails and to verify linking packages to emails. The pattern of `if` conditions suggests that there isn't much in common either - I think it is best to separate these out.

rohitpaulk replied:
> I think it is best to separate these out.

i.e. use separate functions (and templates) for verifying an email vs verifying linking an email to a package

rohitpaulk replied:
> Because it's a content spoofing vector. I thought we had a HackerOne ticket about this but it's not jumping out at me on http://inside.gratipay.com/appendices/disclosures.

I think it'd be good to add a comment in the simplate that mentions this

EdOverflow replied:
> Because it's a content spoofing vector. I thought we had a HackerOne ticket about this but it's not jumping out at me on http://inside.gratipay.com/appendices/disclosures.

It was around the time of [**h1:117187**](https://hackerone.com/reports/117187) and [h1:156542](https://hackerone.com/reports/156542).

whit537 replied:
Good find, @EdOverflow.

> I think it'd be good to add a comment in the simplate that mentions this

Comment in c3677d4943a4be57dd3f4ecb0aee16187f3c30c7, with link to H1.

I've also kicked off disclosure for those two. Seems they fell through the cracks. ☺️

whit537 replied:
> This is confusing 😕 

D'oh! 😞

> `verification.spt` is used both to verify emails and to verify linking packages to emails.

Yeah, it seemed to me that it was best to have one code path for both, since we want linking packages to emails to _also_ verify the email. That is, we don't want to make the user go through two verification flows (one for the address itself, a second for the packages) when from their point of view one should be sufficient. Am I thinking about that right?

> The pattern of `if` conditions suggests that there isn't much in common either

Are you referring to the `if action == 'start-verification'` block? That's the extra bit for the package claiming case, where we load the packages in question. The flows rejoin just after, in the call to [`start_email_verification`](https://github.com/gratipay/gratipay.com/blob/c3677d4943a4be57dd3f4ecb0aee16187f3c30c7/gratipay/models/participant/email.py#L57-L100), which is where the bulk of the common wiring lives.

> I think it is best to separate these out.
> i.e. use separate functions (and templates) for verifying an email vs verifying linking an email to a package

I agree it's pretty dense (especially when we get to [`finish_email_verification`](https://github.com/gratipay/gratipay.com/blob/c3677d4943a4be57dd3f4ecb0aee16187f3c30c7/gratipay/models/participant/email.py#L218-L254)), but I think it's necessary to support the simpler, single-step user experience during package claiming. Do you see a way to refactor while preserving the single-step flow?

rohitpaulk replied:
> Yeah, it seemed to me that it was best to have one code path for both, since we want linking packages to emails to also verify the email. That is, we don't want to make the user go through two verification flows (one for the address itself, a second for the packages) when from their point of view one should be sufficient. Am I thinking about that right?

I see how this makes sense. In that case, do we absolutely *need* to re-verify emails when claiming packages? If a user has a verified email, can't we just allow them to claim the package directly?

whit537 replied:
Decided in [slack](https://gratipay.slackarchive.io/gratipay/page-100/ts-1497614829276345) to reticket dropping the re-verification.

rohitpaulk replied:
^ Discussed on [Slack](https://gratipay.slackarchive.io/gratipay/page-100/ts-1497615566425964).

whit537 replied:
Reticketed as #4520.

whit537 commented:
Rebased, was abfef42b687522c60b4061c6fa362fa6266e9bc0.
